pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -0,0 +1,303 @@
+		if current, err := strconv.ParseBool(value); err == nil && current == desired {
+			return false
+		}
+	}
+
+	return true
+}
+
+// setAutoUpdateAnnotation sets the auto-update annotation key to the specified value.
+func setAutoUpdateAnnotation(accessor metav1.Object, autoUpdate bool) {
+	if accessor.GetAnnotations() == nil {
+		accessor.SetAnnotations(map[string]string{})
+	}
+
+	accessor.GetAnnotations()[flowcontrolv1beta1.AutoUpdateAnnotationKey] = strconv.FormatBool(autoUpdate)
+}
+
+// ensureConfiguration ensures the boostrap configurationWrapper on the cluster based on the specified strategy.
+func ensureConfiguration(wrapper configurationWrapper, strategy ensureStrategy, bootstrap configurationObject) error {
+	name := bootstrap.GetName()
+	configurationType := strategy.Name()
+
+	current, err := wrapper.Get(bootstrap.GetName())
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to retrieve %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+		}
+
+		if strategy.ShouldCreate() {
+			if _, err := wrapper.Create(bootstrap); err != nil {
+				return fmt.Errorf("cannot create %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+			}
+
+			klog.V(2).InfoS(fmt.Sprintf("Successfully created %s", wrapper.TypeName()), "type", configurationType, "name", name)
+			return nil
+		}
+
+		klog.V(3).InfoS(fmt.Sprintf("Skipping creation of %s", wrapper.TypeName()), "type", configurationType, "name", name)
+		return nil
+	}
+
+	klog.V(5).InfoS(fmt.Sprintf("The %s already exists, checking whether it is up to date", wrapper.TypeName()), "type", configurationType, "name", name)
+	newObject, update, err := strategy.ShouldUpdate(current, bootstrap)
+	if err != nil {
+		return fmt.Errorf("failed to determine whether auto-update is required for %s type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	}
+	if !update {
+		if klog.V(3).Enabled() {
+			// TODO: if we use structured logging here the diff gets escaped and very awkward to read in the log
+			klog.Infof("No update required for the %s type=%s name=%q diff: %s", wrapper.TypeName(), configurationType, name, cmp.Diff(current, bootstrap))
+		}
+		return nil
+	}
+
+	if _, err := wrapper.Update(newObject); err != nil {
+		return fmt.Errorf("failed to update the %s, will retry later type=%s name=%q error=%w", wrapper.TypeName(), configurationType, name, err)
+	}
+
+	klog.V(2).Infof("Updated the %s type=%s name=%q diff: %s", wrapper.TypeName(), configurationType, name, cmp.Diff(current, newObject))
+	return nil
+}
+
+func removeConfiguration(wrapper configurationWrapper, bootstrap configurationObject) error {
+	name := bootstrap.GetName()
+
+	current, err := wrapper.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+
+		return fmt.Errorf("failed to retrieve the %s, will retry later name=%q error=%w", wrapper.TypeName(), name, err)
```

Comment:
hmm - this isn't an error right?

If the object doesn't exist, then we're done and we should return nil.

Reply:
I have a guard, if it is a `IsNotFound` error then the function returns `nil`.
```
		if apierrors.IsNotFound(err) {
			return nil
		}
```

Reply:
OK - I missed that. Sorry.